### PR TITLE
Support for Competing Consumers

### PR DESCRIPTION
**Pull request by klinakuf**
In this PR we add support for the second _service pattern_ in SPD called Competing Consumers Group. This PR comes with several significant changes which are detailed below: 

**Intro**

The CompetingConsumersGroup (read https://www.enterpriseintegrationpatterns.com/patterns/messaging/CompetingConsumers.html) may have different realizations in the Palladio model, we follow the realization in the context of the RemoteMeasuringUse case from MOSAIC. The importance of this pattern for SPD is to showcase that in a single model for scaling, we can cover different forms of interaction, e.g., synchronous load balanced replicas (ServiceGroup), asynchronous competing consuming replicas (CompetingConsumersGroup), and their co-existence in the model. 

**What is the structure of a CompetingConsumersGroup according to the MOSAIC RMUC model?**

The picture below shows the initial structure and the intended replication of a CompetingConsumersGroup. There is one broker from which the unit assemblies deplete the queues, which are modeled as PassiveResources. A new unit with the same depletion (consumption) behavior is added upon replication. Furthermore, the usage model that activates the assemblies for consumption is modified by adding a branch and doubling the population (not shown in the figure). 

Screenshot 2023-08-12 at 15 12 18

**SPD Meta-model changes for Competing Consumers**

* In the metamodel of SPD, I add the possibility of referencing the unit assembly for the CompetingConsumersGroup. In addition, for the QueueLength stimulus, I add the possibility to reference a PassiveResource, which models the queue's state. Through the latter, one can model the behavior that the number of consumers changes based on the state of the consumption queue. 
* In the metamodel of the SPD Semantic, I define the CompetingConsumersCfg as similar to ServiceGroupCfg with the difference of defining the broker assembly and the queues of the unit assembly. In the queues, one defines each queue by name, the passive resource representing it, and the usage scenario activating that particular queue's consumption. Not all information is necessary for the transformation (some is retrieved at transformation time). However, it helps to configure a CompetingConsumersGroup and to make the assumptions more explicit (the one shown in the figure). 

**Transformations of Competing Consumers** 

* This PR also contains the transformation of CompetingConsumersCfg, both top-down and bottom-up. 
* This has been tested with the Remote-Measuring-Use-Case, branch slingshot-cc which also contains a valid configuration. 

**Refactorings**

In addition, I made some refactorings, mainly extracting helper/query methods in their libraries and referencing them in the corresponding scripts and some other smaller changes.

### CODE REVIEW

**Review by kjuli — CHANGES_REQUESTED** [review]
Overall it seems to correctly parse on my machine and it works. Just add a bit of information as a documentation to the CompetingConsumer so we can look it up again when we implement the interpreter for this.

**Comment by stiesssh** [comment]
when i use a policy with an Absolute Adjustment of 3, i get 2 new replicas of the Queue consumer assembly. however, they are both on different resource containers, which surprised me, because i assumed they would be on the same resource containern (until the resource container reached the max number of assemblies per host)
@klinakuf is this a bug or a misunderstanding?

**Review by stiesssh — COMMENTED** [review]
- model changes read reasonable. 
- model creation works just fine. 
- transformations read reasonable, except for the points i marked with comments. iirc, they are all questions. @klinakuf if you dont want to write answers, feel free to come over and talk :D 
- execution as described in the PR works
- attempts to break the transformation : 
  * multiple branches -> fatal, as expected
  *  empty queue attribute -> nothing happens,  actually, what is this attribute good for? apparently it's never accessed? is this the part you (floriment) refered to as "make the assumptions more explicit" in the description of the PR above? however, if find it rather confusing to have fields, where i can fill in any thing and they are just ignored :/  
- attempt to combine CCG / EI / SG TargetGroups with other Configurations, e.g. EI TG with Configuration for EI, CCG and SG: results looked reasonable in general, but i did not check all combinations.   

In addition : i really like the description of the PR. in case you aren't already planning to do that, i'd recommend you save some of the explanations and reuse them in the documentation.

File: bundles/org.palladiosimulator.spd/model/SPD.ecore [hunk 1]
```diff
@@ -46,16 +46,18 @@
...
+      <eStructuralFeatures xsi:type="ecore:EReference" name="unitAssembly" eType="ecore:EClass ../../org.palladiosimulator.pcm/model/pcm.ecore#//core/composition/AssemblyContext">
         <eAnnotations source="http://www.eclipse.org/emf/2002/GenModel">
           <details key="documentation" value="The unitAssembly is used to point to the ServiceGroup in PCM. It is used also for disinguishing between different service groups. A prerequisite is that the unit assembly is already connected in a Service Group structure in PCM. "/>
         </eAnnotations>
       </eStructuralFeatures>
     </eClassifiers>
-    <eClassifiers xsi:type="ecore:EClass" name="CompetingConsumersGroup" eSuperTypes="#//targets/TargetGroup"/>
+    <eClassifiers xsi:type="ecore:EClass" name="CompetingConsumersGroup" eSuperTypes="#//targets/TargetGroup">
+      <eStructuralFeatures xsi:type="ecore:EReference" name="unitAssembly" eType="ecore:EClass ../../org.palladiosimulator.pcm/model/pcm.ecore#//core/composition/AssemblyContext"/>
+    </eClassifiers>
```
[the remark this answers is not in the document]
Comment (klinakuf) [reply bundles/org.palladiosimulator.spd/model/SPD.ecore]: provided additional docu

**Comment by stiesssh** [comment]
> Tested out with AbsoluteAdjustment:3, works fine, two new replicas are allocated on a new resource container. Did you also have some intermediate scalings that modified the configuration and the models?

No i didn't do any other adjustments. but maybe i didnt reset the models correctly, let me just try again...

**Comment by stiesssh** [comment]
good news : step adjustments are now correct, tested for step size 1 and 2, realtive adjustments with a percentage, such that the transformation only ever adds 1 assembly per execution works fine as well.

alright-twice-step

the bad news :  relative adjustment with percentage of 50%, min 1, allocates the assemblies still wrong:
bug-twice-rel

**Comment by stiesssh** [comment]
> This is fine while constraints are checked not to exceed the maximum allocations, and any is selected (right now in the code). In this particular adaptation, the creation of the resource container would occur anyhow while two new replicas are created. With the current impl, two possible states are possible: one 3|1 and 2|2, which is okay. Eventually, both will be filled with three each. Do you see a big issue in this?

huh, indeed, if i keep scaling, the 2|2 combination get filled up to 3|3. tbh, not what i expected xD 
in this case i guess it's fine.

**Review by stiesssh — APPROVED** [review]
the suggestion i made above for refactoring those code duplicates is still standing, but i think that's more of a beautification issue. as the transformation is, in my opinion, functional by now,  i'll grant my approval anyway : D

**Review by kjuli — APPROVED** [review]
lgtm

**Pull request merged**